### PR TITLE
Republish editions when dependent editions change

### PR DESCRIPTION
**Pull request by vinayvinay**
https://trello.com/c/oYy2udbc
Proposal: [How to handle dynamic govspeak elements](https://gov-uk.atlassian.net/wiki/x/n4ASAQ)

following up tracking edition's contact dependencies in https://github.com/alphagov/whitehall/pull/2009, we're doing the same by tracking an edition's dependencies on other editions through use of admin links. the aim is to re-publish dependent editions when **draft** editions they depend on get published.

important to note, the dependency ceases to exists once the dependable edition is published, hence we clean-up those dependencies. once an edition moves to a superseded or unarchived state, we can clean the dependencies, because we don't need to republish such editions.

### CODE REVIEW

File: lib/govspeak/editions_extractor.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
+module Govspeak
+  class EditionsExtractor
```
Comment (heathd) [inline lib/govspeak/editions_extractor.rb]: I feel as though this name doesn't quite capture the purpose of this class.

How about `UnresolvedAdminLinkExtractor`. By `Unresolved` I am trying to indicate that it only returns admin links for editions which have not been published yet, but there could be a better word for it.

Comment (vinayvinay) [inline lib/govspeak/editions_extractor.rb]: @heathd: i've renamed this class to `DependableEditionsExtractor`.. it'll not be appropriate to call it an admin links extractor, because it returns editions. adding dependable to the name indicates that its output contains a specific category of editions. let me know if we can improve it further.

File: test/unit/services/listeners/edition_dependencies_test.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        [Contact:#{contact.id}] or read our [official statement](/government/admin/speeches/#{speech.id})", major_change_published_at: Time.zone.now)
+
+      stub_panopticon_registration(news_article)
+
+      assert Whitehall.edition_services.send(service_name, news_article).perform!
+      assert_equal [contact], news_article.depended_upon_contacts
+      assert_equal [speech], news_article.depended_upon_editions
+    end
+
+    test "Whitehall.edition_services.#{service_name} cleans-up a dependable edition's dependency records" do
```
Comment (heathd) [inline test/unit/services/listeners/edition_dependencies_test.rb]: I think this could be made a bit clearer. In the previous test you are (force)publishing an edition and expecting that its list of dependencies are updated. In this test you are (force)publishing a depended-upon edition and expecting that it is no longer depended upon. I think it would be helpful to make it very clear that the thing which is being published is at a different end of the dependency relationship. 

For example:

```
["", "force"].each do |force| 
  test "#{force} publishing a depended-upon edition removes it as a dependency" do
    service_name = force == "" ? :publisher : :force_publisher
    ...
  end
end
```

Comment (vinayvinay) [inline test/unit/services/listeners/edition_dependencies_test.rb]: i've implemented this suggestion, with a slight change. i think the above example hints that we're handling publishing and force publishing differently, whereas we're not.

File: config/initializers/edition_services.rb [hunk 3]
```diff
@@ -7,7 +7,10 @@
 Whitehall.edition_services.subscribe("unpublish") { |event, edition, options| ServiceListeners::SearchIndexer.new(edition).remove! }
 Whitehall.edition_services.subscribe(/^(force_publish|publish)$/) { |_, edition, _| Whitehall::PublishingApi.publish(edition) }
 Whitehall.edition_services.subscribe(/^(force_publish|publish)$/) { |_, edition, _| EditionDependenciesPopulator.new(edition).populate! }
+Whitehall.edition_services.subscribe(/^(force_publish|publish)$/) { |_, edition, _| edition.republish_dependent_editions }
+Whitehall.edition_services.subscribe(/^(force_publish|publish)$/) { |_, edition, _| edition.destroy_dependency_records }
```
Comment (vinayvinay) [inline config/initializers/edition_services.rb]: @heathd: in response to the conversation over Slack, i've made destruction of dependency records explicit, such that it can be visible from `edition_services.rb`. thanks for pointing that out!

File: lib/dependable.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  included do
+    has_many :records_of_dependent_editions, class_name: 'EditionDependency', as: :dependable, dependent: :destroy
+    has_many :dependent_editions, through: :records_of_dependent_editions, source: :edition
+  end
+
+  def republish_dependent_editions
+    dependent_editions.each { |e| Whitehall::PublishingApi.republish(e) }
+  end
+
+  def destroy_records_of_dependent_editions
```
Comment (heathd) [inline lib/dependable.rb]: I don't think that this method belongs in the `Dependable` module because it is not relevant to all dependable things. It's only relevant to `Editions`. I think it would be more appropriate to put this on the `Edition` class

**Comment by heathd** [comment]
Hi Vinay,

I've spent a bit of time thinking through some edge case scenarios. I was concerned about the fact that the dependency is edition->edition but semantically it's edition->document. So I was thinking through possible issues where the depended-upon thing might be published again. I don't think there are issues there, but it was sufficiently complicated to think through that a test case covering that scenario might be warranted. e.g.

1. E2(published) links to E1(draft)
2. publish E1
   (E2 is republished)
3. redraft(E2) producing E3
4. publish E3
   (E2 is not republished)

I realise that this scenario is partly covered but like I said there are quite a lot of moving parts and it might be valuable to have extra coverage here. I think it could go in `test/unit/services/listeners/edition_dependencies_test.rb`
 
I think there's one edge case around unpublishing however, I don't think this is necessarily a show-stopper. The scenario I'm thinking of is:

1. E2(published) links to E1(draft)
2. publish E1
   (E2 is republished)
3. unpublish(E2) setting E2 status to draft
   (no dependency record is created... BUG!)
4. modify title of E2 causing slug to change
5. publish E2
   (E2 is not republished.. BUG!)

I'm writing this a bit late in the day so there's a chance I've made a mistake. I haven't verified this bug in practice, only theoretically. 

Like I said this is not necessarily a show-stopper as I think the scenario is unlikely and the side effect is only a failure to republish a dependent edition in this edge case.

David

File: lib/dependable.rb (hunk 4 above)
Comment (vinayvinay) [inline lib/dependable.rb]: in the interest of not adding more lines to `Edition`, and following the pattern of concerns: `Edition::Identifiable`, `Edition::Workflow`, `Edition::Translatable` etc. i thought of keeping this method here.

**Comment by vinayvinay** [comment]
@heathd
(1) good point about adding a test to cover publishing of subsequent editions of dependable editions. i've added that: https://github.com/alphagov/whitehall/commit/5e2898cc938b579b4c85125815bf03881cabc250
(2) the second edge case you're describing related to unpublishing is not a concern. at step 2 in your illustration, E1 is published, so dependency of E2 on E1 ceases to exist. think about it. once published, E1 is frozen, so it can't be changed. consequently, unpublishing E2 should not require dependency on E1 to be recorded. so there's no BUG! at step 3.

**Comment by heathd** [comment]
@vinayvinay regarding point (2) I'm afraid I made a mistake in the steps, sorry I was tired.

It should have read:

1. E2(published) links to E1(draft)
2. publish E1 (E2 is republished)
3. unpublish(E1) setting E1 status to draft (no dependency record is created... BUG!)
4. modify title of E1 causing slug to change
5. publish E1 (E2 is not republished.. BUG!)

the problem is that the slug of a document CAN be changed when it is unpublished, so when something is unpublished (in step 3) any editions which link to it (E2 in this case) need to have dependencies recreated.

**Comment by vinayvinay** [comment]
@heathd: now i see... that's a good point! if you think it is worth it, i can cover this scenario, shouldn't take too much effort to do that.

**Comment by danielroseman** [comment]
We should fix this scenario. Rather than deleting the edition dependency row when the edition is published, could we have an "active" boolean column that is flipped to false on publish, and back to true on unpublish?

**Comment by vinayvinay** [comment]
@danielroseman: the other option is to fire one query to select editions that have admin links to the edition being unpublished, and populate dependencies from that. this will make sure the dependencies table holds more valuable information.

**Comment by vinayvinay** [comment]
@danielroseman, @heathd: i've covered the scenario where the depended-upon edition is unpublished and published again by delaying destroying records of dependent editions: https://github.com/alphagov/whitehall/commit/25786559022c96bc056fdb40da5c1ec79d8a8374

earlier, they were being destroyed when the depended-upon edition was published, so that we clean-up as soon as they're no longer required. now, we continue to hold onto these associations till the time the depended-upon edition is superseded. there's no side-effect because we republish dependants
only when the depended-upon edition is published / force-published.

we're avoiding introducing complexity in code to handle this, at the cost of storing a few more records in the db. theoretically, a depended-upon edition remains depended-upon till the time it is superseded (because it can be unpublished anytime), so we're good with that way of thinking as well.

**Comment by danielroseman** [comment]
This looks very sensible, although I'll wait for David to comment.

It will lead to more than a "few" more records, since most editions don't get superseded at all, but I don't think that is a problem.

**Comment by vinayvinay** [comment]
@danielroseman: i had checked, i should have said "few hundred". in the copy of WH db i've got there are only 287 published editions that contain admin links.

**Comment by heathd** [comment]
Looks good to me, I'm going to go ahead and merge so it's ready to deploy on Monday.
